pkg/node/node.go

```diff
@@ -101,105 +103,261 @@ func createAddressFilter(
-		keepAddress.String(),
-		serializedPublicKey,
-	)
-
-	return signer, nil
+	return nil
 }
 
 // CalculateSignature calculates a signature over a digest with threshold
 // signer and publishes the result to the keep associated with the signer.
-// In case of failure on signature submission we need to check if the keep is
-// still waiting for the signature. It is possible that other member was faster
-// than the current one and submitted the signature first.
+//
+// The attempt for generating and publishing signature is retried on failure
+// until the provided context is done.
 func (n *Node) CalculateSignature(
 	ctx context.Context,
 	signer *tss.ThresholdSigner,
 	digest [32]byte,
 ) error {
-	signature, err := signer.CalculateSignature(ctx, digest[:], n.networkProvider)
-	if err != nil {
-		return fmt.Errorf("failed to calculate signature: [%v]", err)
-	}
+	keepAddress := common.HexToAddress(signer.GroupID())
 
-	logger.Debugf(
-		"signature calculated:\nr: [%#x]\ns: [%#x]\nrecovery ID: [%d]\n",
-		signature.R,
-		signature.S,
-		signature.RecoveryID,
-	)
+	attemptCounter := 0
+	for {
+		attemptCounter++
 
-	keepAddress := common.HexToAddress(signer.GroupID())
+		logger.Infof(
+			"calculate signature for keep [%s]; attempt [%v]",
+			keepAddress.String(),
+			attemptCounter,
+		)
+
+		// Global timeout for generating a signature exceeded.
+		// We are giving up and leaving this function.
+		if ctx.Err() != nil {
+			return fmt.Errorf("signing timeout exceeded")
+		}
+
+		// Calculate the signature executing threshold signing protocol with
+		// other keep members.
+		//
+		// If threshold signing fails, we retry from the beginning.
+		signature, err := signer.CalculateSignature(ctx, digest[:], n.networkProvider)
+		if err != nil {
+			logger.Errorf("failed to calculate signature: [%v]", err)
+			continue
+		}
+
+		logger.Debugf(
+			"signature calculated:\nr: [%#x]\ns: [%#x]\nrecovery ID: [%d]\n",
+			signature.R,
+			signature.S,
+			signature.RecoveryID,
+		)
+
+		// We have the signature so now we need to publish it.
+		// This function implements internal retries so we do not need to
+		// retry here.
+		return n.publishSignature(ctx, keepAddress, digest, signature)
+	}
+}
 
-	if err := n.ethereumChain.SubmitSignature(keepAddress, signature); err != nil {
+// publisSignature takes the provided signature and attempts to publish it to
+// the chain. It implements retry mechanism allowing to attempt to publish again
+// in case of a failure.
+//
+// We do implement a retry in this function because the retry mechanism is much
+// more complex than in case of e.g. publishSignerPublicKey. Although all keep
+// members are supposed to try to publish the signature, only one transaction
+// succeeds. For each attempt, we need to check if the keep still awaits
+// a signature. Also, we need to implement some sane delay between attempts so
+// that we do not waste gas.
+func (n *Node) publishSignature(
+	ctx context.Context,
```

Comment:
We're not using `ctx` in this function, so can remove this parameter.
On the other hand maybe we want to use this to check in the retry loop that we reached the timeout?

Reply:
> On the other hand maybe we want to use this to check in the retry loop that we reached the timeout?

We do, good catch. Fixed.